server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java

```diff
@@ -84,22 +97,116 @@ public Disk getDiskMetadata() {
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(diskMetadata);
+        return Objects.hash(diskMetadata, shardLimitsMetadata);
     }
 
     @Override
     public String toString() {
-        return "HealthMetadata{diskMetadata=" + Strings.toString(diskMetadata) + '}';
+        return "HealthMetadata{diskMetadata=" + Strings.toString(diskMetadata) + ", shardLimitsMetadata=" + shardLimitsMetadata + "}";
+    }
+
+    /**
+     * Contains the thresholds needed to determine the health of a cluster when it comes to the amount of room available to create new
+     * shards. These values are determined by the elected master.
+     */
+    public record ShardLimits(int maxShardsPerNode, int maxShardsPerNodeFrozen) implements ToXContentFragment, Writeable {
+
+        private static final String TYPE = "shard_limits";
+        private static final ParseField MAX_SHARDS_PER_NODE = new ParseField("max_shards_per_node");
+        private static final ParseField MAX_SHARDS_PER_NODE_FROZEN = new ParseField("max_shards_per_node_frozen");
+        static final TransportVersion VERSION_SUPPORTING_SHARD_LIMIT_FIELDS = TransportVersion.V_8_8_0;
+
+        public ShardLimits {
+            if (maxShardsPerNode < 1) {
+                throw new IllegalArgumentException(
+                    "Setting "
+                        + ShardLimitValidator.SETTING_CLUSTER_MAX_SHARDS_PER_NODE.getKey()
+                        + " must have a value greater than 1 but it was ["
+                        + maxShardsPerNode
+                        + "]"
+                );
+            }
+
+            if (maxShardsPerNodeFrozen < 1) {
+                throw new IllegalArgumentException(
+                    "Setting "
+                        + ShardLimitValidator.SETTING_CLUSTER_MAX_SHARDS_PER_NODE_FROZEN.getKey()
+                        + " must have a value greater than 1 but it was ["
+                        + maxShardsPerNodeFrozen
+                        + "]"
+                );
+            }
+        }
+
+        static ShardLimits readFrom(StreamInput in) throws IOException {
+            return new ShardLimits(in.readInt(), in.readInt());
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(MAX_SHARDS_PER_NODE.getPreferredName(), maxShardsPerNode);
+            builder.field(MAX_SHARDS_PER_NODE_FROZEN.getPreferredName(), maxShardsPerNodeFrozen);
+            return builder;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeInt(maxShardsPerNode);
+            out.writeInt(maxShardsPerNodeFrozen);
+        }
+
+        public boolean isUnknown() {
+            return maxShardsPerNode != -1 && maxShardsPerNodeFrozen != -1;
+        }
```

Comment:
I believe this shouldn't be needed anymore? A missing (`null`) ShardLimits represents `unknown`